pkg/kubelet/kuberuntime/kuberuntime_manager.go

```diff
@@ -0,0 +1,1150 @@
+func (m *kubeGenericRuntimeManager) Type() string {
+	return m.runtimeType
+}
+
+// runtimeVersion implements kubecontainer.Version interface by implementing
+// Compare() and String()
+type runtimeVersion struct {
+	*semver.Version
+}
+
+func newRuntimeVersion(version string) (runtimeVersion, error) {
+	sem, err := semver.NewVersion(version)
+	if err != nil {
+		return runtimeVersion{}, err
+	}
+	return runtimeVersion{sem}, nil
+}
+
+func (r runtimeVersion) Compare(other string) (int, error) {
+	v, err := semver.NewVersion(other)
+	if err != nil {
+		return -1, err
+	}
+
+	if r.LessThan(*v) {
+		return -1, nil
+	}
+	if v.LessThan(*r.Version) {
+		return 1, nil
+	}
+	return 0, nil
+}
+
+// Version returns the version information of the container runtime.
+func (m *kubeGenericRuntimeManager) Version() (kubecontainer.Version, error) {
+	typedVersion, err := m.runtimeService.Version(kubeRuntimeAPIVersion)
+	if err != nil {
+		glog.Errorf("Get remote runtime version failed: %v", err)
+		return nil, err
+	}
+
+	return newRuntimeVersion(typedVersion.GetVersion())
+}
+
+// APIVersion returns the cached API version information of the container
+// runtime. Implementation is expected to update this cache periodically.
+// This may be different from the runtime engine's version.
+func (m *kubeGenericRuntimeManager) APIVersion() (kubecontainer.Version, error) {
+	typedVersion, err := m.runtimeService.Version(kubeRuntimeAPIVersion)
+	if err != nil {
+		glog.Errorf("Get remote runtime version failed: %v", err)
+		return nil, err
+	}
+
+	return newRuntimeVersion(typedVersion.GetRuntimeApiVersion())
+}
+
+// Status returns error if the runtime is unhealthy; nil otherwise.
+func (m *kubeGenericRuntimeManager) Status() error {
+	_, err := m.runtimeService.Version(kubeRuntimeAPIVersion)
+	if err != nil {
+		glog.Errorf("Checkout remote runtime status failed: %v", err)
+		return err
+	}
+
+	return nil
+}
+
+// GetPods returns a list of containers grouped by pods. The boolean parameter
+// specifies whether the runtime returns all containers including those already
+// exited and dead containers (used for garbage collection).
+func (m *kubeGenericRuntimeManager) GetPods(all bool) ([]*kubecontainer.Pod, error) {
+	pods := make(map[kubetypes.UID]*kubecontainer.Pod)
+	var result []*kubecontainer.Pod
+
+	sandboxes, err := m.getKubeletPodSandboxes(all)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, s := range sandboxes {
+		name, namespace, _ := kubecontainer.ParsePodFullName(s.GetName())
+		labelledPodInfo := GetPodSandboxInfoFromLabels(s.Labels)
+		pods[labelledPodInfo.PodUID] = &kubecontainer.Pod{
+			ID:        labelledPodInfo.PodUID,
+			Name:      name,
+			Namespace: namespace,
+		}
+	}
+
+	containers, err := m.getKubeletContainers(all)
+	if err != nil {
+		return nil, err
+	}
+	for _, c := range containers {
+		labelledInfo := GetContainerInfoFromLabels(c.Labels)
+		pod, found := pods[labelledInfo.PodUID]
+		if !found {
```

Comment:
This could cause trouble. In docker's case, if the infra container died unexpectedly, kubelet still want to see and handle containers in the pod -- mostly to get the status and kill them. Can we still get the pod UID from the container labels and fix this?